zircon-object/src/vm/stream.rs

```diff
@@ -0,0 +1,265 @@
+        let io_vec = unsafe { slice::from_raw_parts(data.as_ptr(), length) };
+        self.vmo.write(offset, io_vec)?;
+        Ok(length)
+    }
+
+    /// Write data to the stream at a given offset
+    #[allow(unsafe_code)]
+    pub fn write_at(&self, data: &[u8], offset: usize) -> ZxResult<usize> {
+        let count = data.len();
+        let mut content_size = self.vmo.content_size();
+        let (target_size, overflow) = offset.overflowing_add(count);
+        if overflow {
+            return Err(ZxError::FILE_BIG);
+        }
+        if target_size > content_size {
+            content_size = self.vmo.set_content_size_and_resize(target_size, offset)?;
+        }
+        if offset >= content_size {
+            return Err(ZxError::NO_SPACE);
+        }
+        let length = count.min(content_size - offset);
+        self.vmo.write(offset, data)?;
+        // unsafe code can be avoided by adding more interface to VmObject and HAL
+        let io_vec = unsafe { slice::from_raw_parts(data.as_ptr(), length) };
+        self.vmo.write(offset, io_vec)?;
+        Ok(length)
+    }
+
+    /// Modify the current seek offset of the stream
+    pub fn seek(&self, seek_origin: SeekOrigin, offset: isize) -> ZxResult<usize> {
+        let mut seek = self.seek.lock();
+        match seek_origin {
+            SeekOrigin::SeekOriginStart => {
+                if offset < 0 {
+                    return Err(ZxError::INVALID_ARGS);
+                }
+                *seek = offset as usize;
+            }
+            SeekOrigin::SeekOriginCurrent => {
+                if offset >= 0 {
+                    let (target, overflow) = seek.overflowing_add(offset as usize);
+                    if overflow {
+                        return Err(ZxError::INVALID_ARGS);
+                    }
+                    *seek = target;
+                } else {
+                    let target = *seek as isize + offset;
+                    if *seek as isize > 0 && target < 0 {
+                        return Err(ZxError::INVALID_ARGS);
+                    }
+                    *seek = target as usize;
+                }
+            }
+            SeekOrigin::SeekOriginEnd => {
+                let content_size = self.vmo.content_size();
+                if offset >= 0 {
+                    let (target, overflow) = content_size.overflowing_add(offset as usize);
+                    if overflow {
+                        return Err(ZxError::INVALID_ARGS);
+                    }
+                    *seek = target;
+                } else {
+                    let target = content_size as isize + offset;
+                    if content_size as isize > 0 && target < 0 {
+                        return Err(ZxError::INVALID_ARGS);
+                    }
+                    *seek = target as usize;
+                }
+            }
+            _ => unreachable!(),
+        }
+        Ok(*seek)
```

Comment:
Is this ok?
```suggestion
        let mut seek = self.seek.lock();
        let origin = match seek_origin {
            SeekOrigin::Start => 0,
            SeekOrigin::Current => *seek,
            SeekOrigin::End => self.vmo.content_size(),
        };
        let (target, overflow) = (origin as isize).overflowing_add(offset);
        if overflow || target < 0 {
            return Err(ZxError::INVALID_ARGS);
        }
        *seek = target as usize;
        Ok(*seek)
```

Reply:
This one is ok.

```rust
let mut seek = self.seek.lock();
let origin: usize = match seek_origin {
    SeekOrigin::Start => 0,
    SeekOrigin::Current => *seek,
    SeekOrigin::End => self.vmo.content_size(),
};
if offset >= 0 {
    let (target, overflow) = origin.overflowing_add(offset as usize);
    if overflow {
        return Err(ZxError::INVALID_ARGS);
    }
    *seek = target;
} else {
    let target = origin as isize + offset;
    if origin as isize >= 0 && target < 0 {
        return Err(ZxError::INVALID_ARGS);
    }
    *seek = target as usize;
}
Ok(*seek)
```

The code suggested above will fail when :
* origin > MAX_ISIZE and offset < 0, which is correct but end up with target < 0
* origin > MAX_ISIZE and origin + offset > MAX_USIZE, which overflows but end up with correct

Reply:
😵